src/unix_socket.rs

```diff
@@ -0,0 +1,168 @@
+use std::future::Future;
+use std::path::PathBuf;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use anyhow::{anyhow, Result};
+use pin_project_lite::pin_project;
+use reqwest::blocking::{Request, Response};
+use reqwest::header::{HeaderValue, HOST};
+
+pub struct UnixClient {
+    rt: tokio::runtime::Runtime,
+    socket_path: PathBuf,
+    timeout: Option<Duration>,
+}
+
+impl UnixClient {
+    pub fn new(socket_path: PathBuf, timeout: Option<Duration>) -> Self {
+        let rt = tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+
+        Self {
+            rt,
+            socket_path,
+            timeout,
+        }
+    }
+
+    async fn connect(&self) -> Result<hyper::client::conn::http1::SendRequest<reqwest::Body>> {
+        // TODO: Add support for Windows named pipes by replacing UnixStream with namedPipeClient.
+        // See https://docs.rs/tokio/latest/tokio/net/windows/named_pipe/struct.ClientOptions.html#method.open
+        let stream = tokio::net::UnixStream::connect(&self.socket_path).await?;
+        let (sender, conn) = hyper::client::conn::http1::Builder::new()
+            .title_case_headers(true)
+            .handshake(hyper_util::rt::TokioIo::new(stream))
+            .await?;
+
+        tokio::task::spawn(async move {
+            if let Err(err) = conn.await {
+                log::error!("Connection failed: {:?}", err);
+            }
+        });
+
+        Ok(sender)
+    }
+
+    pub fn execute(&self, request: Request) -> Result<Response> {
+        self.rt.block_on(async {
+            let http_request = into_async_request(request)?;
+
+            let mut sender = with_timeout(self.connect(), self.timeout).await??;
+            let response = with_timeout(sender.send_request(http_request), self.timeout).await??;
```

Comment:
Does this mean we have a separate timeout for each stage?

Will there be a timeout for e.g. a large file upload?

Reply:
Yup, this is too aggressive. Your streaming example server times out and that kind of thing doesn't happen over TCP.

The [docs](https://docs.rs/reqwest/latest/reqwest/blocking/struct.ClientBuilder.html#method.timeout) say the timeout applies to individual connect/read/write operations. So I think that if you have a timeout of 1 second, as long as each individual read/write takes less than 1 second, it can go on indefinitely.

I don't know how easy it is to match that precisely but putting the timeout just on the `connect` step would already be decent.